http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/MultiplexedChannelRecord.java

```diff
@@ -31,165 +27,194 @@
-        this.availableStreams = new AtomicLong(maxConcurrencyPerConnection);
-        this.childChannels = new ConcurrentHashMap<>(saturatedCast(maxConcurrencyPerConnection));
-        this.channelReleaser = channelReleaser;
-    }
+    // Only read or write in the connection.eventLoop()
+    private final Map<ChannelId, Http2StreamChannel> childChannels = new HashMap<>();
+
+    // Only write in the connection.eventLoop()
+    private volatile RecordState state = RecordState.OPEN;
 
-    @SdkTestInternalApi
-    MultiplexedChannelRecord(Future<Channel> connectionFuture,
-                             Channel connection,
-                             long maxConcurrencyPerConnection,
-                             BiConsumer<Channel, MultiplexedChannelRecord> channelReleaser) {
-        this.connectionFuture = connectionFuture;
-        this.childChannels = new ConcurrentHashMap<>(saturatedCast(maxConcurrencyPerConnection));
-        this.availableStreams = new AtomicLong(maxConcurrencyPerConnection);
-        this.channelReleaser = channelReleaser;
+    MultiplexedChannelRecord(Channel connection, long maxConcurrencyPerConnection) {
         this.connection = connection;
+        this.maxConcurrencyPerConnection = maxConcurrencyPerConnection;
+        this.availableChildChannels = new AtomicLong(maxConcurrencyPerConnection);
+    }
+
+    boolean acquireStream(Promise<Channel> promise) {
+        if (reserveStream()) {
+            releaseReservationOnFailure(promise);
+            acquireReservedStream(promise);
+            return true;
+        }
+        return false;
     }
 
-    MultiplexedChannelRecord acquire(Promise<Channel> channelPromise) {
-        availableStreams.decrementAndGet();
-        if (connection != null) {
-            createChildChannel(channelPromise);
-        } else {
-            connectionFuture.addListener((GenericFutureListener<Future<Channel>>) future -> {
-                if (future.isSuccess()) {
-                    connection = future.getNow();
-                    connection.attr(CHANNEL_POOL_RECORD).set(this);
-                    createChildChannel(channelPromise);
-                } else {
-                    channelPromise.setFailure(future.cause());
-                    channelReleaser.accept(connection, this);
+    private void acquireReservedStream(Promise<Channel> promise) {
+        doInEventLoop(connection.eventLoop(), () -> {
+            if (state != RecordState.OPEN) {
+                String message = "Connection received GOAWAY or was closed while acquiring new stream.";
+                promise.setFailure(new IllegalStateException(message));
+                return;
+            }
+
+            Future<Http2StreamChannel> streamFuture = new Http2StreamChannelBootstrap(connection).open();
+            streamFuture.addListener((GenericFutureListener<Future<Http2StreamChannel>>) future -> {
+                warnIfNotInEventLoop(connection.eventLoop());
```

Comment:
Curious how this would happen in the given code

Reply:
I have a mistrust of Netty's HTTP2 implementation. I'll remove it if you think it's unlikely.

Reply:
Haha got it. I think it's fine to leave it, but it's probably unlikely to happen as long as `Http2StreamChannelBootstrap` does the obvious thing and uses the `connection` to create the returned future but you never know.